crates/ra_hir/src/ty/tests/never_type.rs

```diff
@@ -0,0 +1,258 @@
+use super::type_at;
+
+#[test]
+fn infer_never1() {
+    let t = type_at(
+        r#"
+//- /main.rs
+fn test() {
+    let t = return;
+    t<|>;
+}
+"#,
+    );
+    assert_eq!(t, "!");
+}
+
+#[test]
+fn infer_never2() {
+    let t = type_at(
+        r#"
+//- /main.rs
+trait Foo { fn gen() -> Self; }
+impl Foo for ! { fn gen() -> Self { loop {} } }
+impl Foo for () { fn gen() -> Self { loop {} } }
+
+fn test() {
+    let a = Foo::gen();
```

Comment:
we don't resolve trait associated method calls yet. so the reason this test works is that `a` simply gets unknown type here, so it could be simplified ;)